tests/test_erc721.py

```diff
@@ -61,56 +63,182 @@ def test_isApprovedForAll(c, w3):
 
 
 def test_transferFrom(c, w3, assert_tx_failed, get_logs):
-    owner, operator, someone = w3.eth.accounts[:3]
+    someone, operator = w3.eth.accounts[1:3]
 
     # transfer from zero address
     assert_tx_failed(lambda: c.transferFrom(
-        ZERO_ADDRESS, operator, firstTokenId, transact={'from': someone}))
+        ZERO_ADDRESS, operator, SOMEONE_TOKEN_IDS[0], transact={'from': someone}))
 
     # transfer to zero address
     assert_tx_failed(lambda: c.transferFrom(
-        someone, ZERO_ADDRESS, firstTokenId, transact={'from': someone}))
+        someone, ZERO_ADDRESS, SOMEONE_TOKEN_IDS[0], transact={'from': someone}))
 
     # transfer token without ownership
     assert_tx_failed(lambda: c.transferFrom(
-        someone, operator, secondTokenId, transact={'from': someone}))
+        someone, operator, OPERATOR_TOKEN_ID, transact={'from': someone}))
+
+    # transfer invalid token
+    assert_tx_failed(lambda: c.transferFrom(
+        someone, operator, INVALID_TOKEN_ID, transact={'from': someone}))
+
+    # transfer by owner
+    tx_hash = c.transferFrom(
+        someone, operator, SOMEONE_TOKEN_IDS[0], transact={'from': someone})
 
+    logs = get_logs(tx_hash, c, 'Transfer')
+
+    assert len(logs) > 0
+    args = logs[0].args
+    assert args._from == someone
+    assert args._to == operator
+    assert args._tokenId == SOMEONE_TOKEN_IDS[0]
+    assert c.ownerOf(SOMEONE_TOKEN_IDS[0]) == operator
+    assert c.balanceOf(someone) == 2
+    assert c.balanceOf(operator) == 2
+
+    # transfer by approved
+    c.approve(operator, SOMEONE_TOKEN_IDS[1], transact={'from': someone})
     tx_hash = c.transferFrom(
-        someone, operator, firstTokenId, transact={'from': someone})
+        someone, operator, SOMEONE_TOKEN_IDS[1], transact={'from': someone})
+
+    logs = get_logs(tx_hash, c, 'Transfer')
+
+    assert len(logs) > 0
+    args = logs[0].args
+    assert args._from == someone
+    assert args._to == operator
+    assert args._tokenId == SOMEONE_TOKEN_IDS[1]
+    assert c.ownerOf(SOMEONE_TOKEN_IDS[1]) == operator
+    assert c.balanceOf(someone) == 1
+    assert c.balanceOf(operator) == 3
+
+    # transfer by operator
```

Comment:
For readability and maintainability, it is better to devide the case of `transfer by owner` and of `transfer by approved` into different functions.